src/openeo_aggregator/backend.py

```diff
@@ -620,6 +620,190 @@ def get_log_entries(self, job_id: str, user_id: str, offset: Optional[str] = Non
+            backends: MultiBackendConnection,
+            processing: AggregatorProcessing
+    ):
+        super(AggregatorSecondaryServices, self).__init__()
+        self._backends = backends
+        self._processing = processing
+
+    def service_types(self) -> dict:
+        """https://openeo.org/documentation/1.0/developers/api/reference.html#operation/list-service-types"""
+
+        service_types = {}
+
+        def merge(formats: dict, to_add: dict):
+            for name, data in to_add.items():
+                if name.lower() not in {k.lower() for k in formats.keys()}:
+                    formats[name] = data
+
+        # Collect all service types from the backends.
+        for con in self._backends:
+            try:
+                types_to_add = con.get("/service_types").json()
+            except Exception as e:
+                # TODO: fail instead of warn?
+                _log.warning(f"Failed to get service_types from {con.id}: {e!r}", exc_info=True)
+                continue
+            # TODO #1 smarter merging:  parameter differences?
+            merge(service_types, types_to_add)
+
+        return service_types
+
+    def list_services(self, user_id: str) -> List[ServiceMetadata]:
+        """https://openeo.org/documentation/1.0/developers/api/reference.html#operation/list-services"""
+        # TODO: user_id is not used, how to authenticate when we use the BackendConnection?
+
+        all_services = []
+        def merge(services, to_add):
+            # For now ignore the links
+            services_to_add = to_add.get("services")
+            if services_to_add:
+                services_metadata = [ServiceMetadata.from_dict(s) for s in services_to_add]
+                services.extend(services_metadata)
+
+        # Collect all services from the backends.
+        for con in self._backends:
+            with con.authenticated_from_request(request=flask.request, user=User(user_id)):
+                services_json = None
+                try:
+                    services_json = con.get("/services").json()
+                except Exception as e:
+                    _log.warning(f"Failed to get services from {con.id}: {e!r}", exc_info=True)
+                    continue
+
+                if services_json:
+                    merge(all_services, services_json)
+
+        return all_services
+
+    def service_info(self, user_id: str, service_id: str) -> ServiceMetadata:
+        """https://openeo.org/documentation/1.0/developers/api/reference.html#operation/describe-service"""
+        # TODO: user_id is not used, how to authenticate when we use the BackendConnection?
+
+        # TODO: can there ever be a service with the same ID in multiple back-ends? (For the same user)
+        for con in self._backends:
+            with con.authenticated_from_request(request=flask.request, user=User(user_id)):
+                try:
+                    service_json = con.get(f"/services/{service_id}").json()
```

Comment:
I agree. Querying all backends just to be able to do another operation is not very efficient and we can easily avoid it. That is a lot of extra HTTP roundtrips that are really unnecessary, even if it could be cached.

Solved by adding the class `ServiceIdMapping`  and  the method `AggregatorSecondaryServices._get_connection_and_backend_service_id` which is used to retrieve the right backend connection where needed.